cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,950 @@
+        // use +MAXINDEX to optimize the host constexpr variable `MAXINDEX` to be used in the device function
+        return (::cuda::std::min)((MAX_EXP - exp) / BIN_WIDTH, +MAXINDEX);
+      }
+    }
+    return ((MAX_EXP + EXP_BIAS) - exp) / BIN_WIDTH;
+  }
+
+  /// Get index of manually specified binned double precision
+  /// The index of a binned type is the bin that it corresponds to. Higher
+  /// indices correspond to smaller bins.
+  _CCCL_DEVICE _CCCL_FORCEINLINE int binned_index() const
+  {
+    return ((MAX_EXP + MANT_DIG - BIN_WIDTH + 1 + EXP_BIAS) - exp_val(primary(0))) / BIN_WIDTH;
+  }
+
+  /// Check if index of manually specified binned floating-point is 0
+  /// A quick check to determine if the index is 0
+  _CCCL_DEVICE _CCCL_FORCEINLINE bool binned_index0() const
+  {
+    return exp_val(primary(0)) == MAX_EXP + EXP_BIAS;
+  }
+
+  //! Update manually specified binned fp with a scalar (X -> Y)
+  //!
+  //! This method updates the binned fp to an index suitable for adding numbers
+  //! with absolute value less than @p max_abs_val
+  //!
+  //!@param incpriY stride within Y's primary vector (use every incpriY'th element)
+  //!@param inccarY stride within Y's carry vector (use every inccarY'th element)
+  _CCCL_DEVICE void binned_dmdupdate(const ftype max_abs_val, const int incpriY, const int inccarY)
+  {
+    if (is_nan_inf_v(primary(0)))
+    {
+      return;
+    }
+
+    int X_index = binned_dindex(max_abs_val);
+    if (is_zero_v(primary(0)))
+    {
+      _CCCL_PRAGMA_UNROLL_FULL()
+      for (int i = 0; i < FOLD; i++)
+      {
+        primary(i * incpriY) = binned_bins(i + X_index);
+        carry(i * inccarY)   = 0.0;
+      }
+    }
+    else
+    {
+      int shift = binned_index() - X_index;
+      if (shift > 0)
+      {
+        _CCCL_PRAGMA_UNROLL_FULL()
+        for (int i = FOLD - 1; i >= 1; i--)
+        {
+          if (i < shift)
+          {
+            break;
+          }
+          primary(i * incpriY) = primary((i - shift) * incpriY);
+          carry(i * inccarY)   = carry((i - shift) * inccarY);
+        }
+        _CCCL_PRAGMA_UNROLL_FULL()
+        for (int j = 0; j < FOLD; j++)
+        {
+          if (j >= shift)
+          {
+            break;
+          }
+          primary(j * incpriY) = binned_bins(j + X_index);
+          carry(j * inccarY)   = 0.0;
+        }
+      }
+    }
+  }
+
+  //! Add scalar @p X to suitably binned manually specified binned fp (Y += X)
+  //!
+  //! Performs the operation Y += X on an binned type Y where the index of Y is
+  //! larger than the index of @p X
+  //!
+  //!@param incpriY stride within Y's primary vector (use every incpriY'th element)
+  _CCCL_DEVICE void binned_dmddeposit(const ftype X, const int incpriY)
+  {
+    ftype M;
+    ftype x = X;
+
+    if (is_nan_inf_v(x) || is_nan_inf_v(primary(0)))
+    {
+      primary(0) += x;
+      return;
+    }
+
+    if (binned_index0())
+    {
+      M        = primary(0);
```

Comment:
Please avoid one letter variable names. More expressive names mean less bugs and less confusion for any reader

Reply:
This is an implementation detail, and is entirely derived from [this repo](https://github.com/maddyscientist/reproducible_floating_sums/blob/feature/cuda/src/reproducible_floating_accumulator.hpp), and this is totally hidden from user. The repo's functions are implemented from this [paper](https://willowahrens.io/assets/documents/ahrens_algorithms_2020.pdf), and its quite hard to correlate the variable names with the equations. In my opinion its better to leave it as is, as it will be easier to compare with the reference implementation.